### ISSUE TITLE
How is this different from Nebula Mesh VPN?

### ISSUE THREAD

**Issue by lonnietc** [yggdrasil-network/yggdrasil-go#866]
Hello,

I am researching Golang Mesh topologies and designs for which I find Yggdrasil to be very interesting and seems to have a lot of promise but am a bit confused on how this is really different from other Mesh solutions like, for example:

https://github.com/slackhq/nebula

The above on is a bit more mature in development and put out by the Slack software developers which seems to run pretty good as well, but I am trying to determine what makes Yggdrasil different as it may be addressing a few items that Nebula Mesh VPN does not.

Any input on this would be greatly appreciated.

**Comment by Arceliar** [comment]
Disclaimer, I haven't used nebula or looked into it much, so take what I say about how it works with a grain of salt.

As far as I understand, when nebula uses the word mesh, they mean a [fully connected network](https://en.wikipedia.org/wiki/Network_topology#Fully_connected_network). I would use nebula in cases where I have two or more nodes that are connected to the same network (e.g. the internet), and I want to establish secure tunnels between them. Nebula does not appear to do any routing, as far as I can tell, so it's always a direct connection (maybe with some single relay mechanism if NAT hole punching fails? That wasn't clear to me from a quick skim through the documentation). It's also very much a product that the company is shipping, so it's supposed to be useful and user-friendly.

When we use the word mesh to talk about yggdrasil, we mean the [mesh networking](https://en.wikipedia.org/wiki/Mesh_networking) sense of the word. It's meant for cases where nodes are *not* all reachable over the same network, and so they need to route through each other. Aside from that, yggdrasil is a research project. The goal here is not to ship a usable product, it's to do R&D, test, and (eventually) document *how* to do scalable routing. Specifically, it's a sort-of continuation of the academic work on compact routing described in e.g. [arxiv:0708.2309](https://arxiv.org/abs/0708.2309). Since we're not trying to ship a product, the project deliberately does not do certain things that would be convenient for the user, but would get in the way of research (e.g. it doesn't automatically find and connect to nodes over the internet, you need to configure internet peers manually).

In short: as far as I can tell, nebula doesn't do routing. In yggdarsil, routing is the whole point. If nebula works for your use case, then using yggdrasil instead is almost certainly a bad idea. On the other hand, if what you care about is yggdrasil's use case, then using nebula isn't an option at all.

**Comment by lonnietc** [comment]
Hello @Arceliar and Thanks for your wonderful and detailed description of yggdrasil.  

Your points clear things up for me and I am interested to explore yggdrasil more to also see if it can be applied to some massively scalable P2P projects that I am currently working towards as I think that the techniques, and approach that yggdrasil uses is very interesting and could offer significant advantages from a Mesh routing prospective. 

I was able to install it and and configure a geographically close to me Peer from:

https://github.com/yggdrasil-network/public-peers

after which I was able navigate to a number of your listed services page:

https://yggdrasil-network.github.io/services.html

to get a feel for how it navigates.

While I was able to navigate to a couple of places, it did seem as though there was a bit more latency than going to some sites without using Yggdrasil, but I also understand that there is some overheard due to encryption, etc, and that this was by no means any type of benchmarking at all.

I want to continue to investigate Yggdrasil much more to see how well it can scale to be sure.

On that note, I also came across the network map (http://51.15.204.214/) and am wondering if this is a dynamic map that updates all of the Yggdrasil Peers often since I really like what you have done to display and interactive map as well.

I think that the project has GREAT potential and am going to try and do more research with it too.
Thanks again

**Comment by neilalexander** [comment]
> While I was able to navigate to a couple of places, it did seem as though there was a bit more latency than going to some sites without using Yggdrasil

This is likely just a side effect of how people are peered with each other on the public network.

For example, if a node in London peers directly to a node in Sydney, and ends up needing to use that peering to reach someone else, that’s a lot of latency accumulated just for that single hop. Now imagine that many nodes on the network are probably peered in equally strange ways — because we can’t stop people from doing that — you’ll get higher latency because traffic is taking routes that probably don’t make sense geographically. (Yggdrasil is not aware of physical geography.)

We don’t have any guarantee that any new user joining the network will actually pick peers that optimise for latency or make any kind of geographic sense. This isn’t strictly an Yggdrasil-specific problem (it would also happen with any other network if there were obviously indirect links) and it wouldn’t really happen if people very consciously only peered with the nodes to which they have the lowest latencies.

However, because we have to effectively test the design of the network as an overlay over the internet like it is today (instead of building lots of dedicated physical links) we don’t really have any alternative.

> but I also understand that there is some overheard due to encryption

Indeed, although the cost of encryption is really only incurred at each end of a given connection. The nodes in between which are carrying the encrypted traffic don’t (and more importantly, *can’t*) decrypt the traffic.
